### PR TITLE
Fix ClassCastException in JwtDecoders fromIssuerLocation

### PR DESCRIPTION
**Pull request by tiborkoch**
Fix for https://github.com/spring-projects/spring-security/issues/9651

### CODE REVIEW

File: oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtDecoderProviderConfigurationUtils.java [hunk 1]
```diff
@@ -90,7 +90,8 @@ static void validateIssuer(Map<String, Object> configuration, String issuer) {
 			List<? extends JWK> jwks = jwkSource.get(new JWKSelector(jwkMatcher), null);
 			for (JWK jwk : jwks) {
 				if (jwk.getAlgorithm() != null) {
-					jwsAlgorithms.add((JWSAlgorithm) jwk.getAlgorithm());
+					JWSAlgorithm jwsAlgorithm = JWSAlgorithm.parse(jwk.getAlgorithm().getName());
```
Comment (jzheaux) [inline oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtDecoderProviderConfigurationUtils.java:93]: Will you please add a test that confirms the fix? It should fail without your change and pass with your change.

Also in that test, would you please prefix it with the issue number like so:

```java
// gh-9651
@Test
public void getSignatureAlgorithmsWhenAlgorithmThenParses() {
 // ...
}
```

**Review by jzheaux — CHANGES_REQUESTED** [review]
Thanks for the PR, @tiborkoch! Please see my review comments inline.

In addition, would you please include the issue number in your squashed Git commit? Something like this:

```
Fix ClassCastException

Closes gh-9651
```

File: oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtDecoderProviderConfigurationUtils.java (hunk 1 above)
Comment (tiborkoch) [reply oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtDecoderProviderConfigurationUtils.java:93]: added a test and updated the commit message

**Comment by jzheaux** [comment]
Thanks, @tiborkoch! This is now merged into `master`.

**Comment by Blockost** [comment]
I'm having the exact same issue but with `spring-security-oauth2-jose@5.6.1`. I set `spring.security.oauth2.resourceserver.jwt.jwk-set-uri` to my Auth0 endpoint but without success.

The line that raises the error is the one at 411 of class  `NimbusReactiveJwtDecoder`:
```java
private JWKSelector createSelector(Function<JWSAlgorithm, Boolean> expectedJwsAlgorithms, Header header) {
	JWSHeader jwsHeader = (JWSHeader) header; // <--- Error thrown here !
	if (!expectedJwsAlgorithms.apply(jwsHeader.getAlgorithm())) {
		throw new BadJwtException("Unsupported algorithm of " + header.getAlgorithm());
	}
	return new JWKSelector(JWKMatcher.forJWSHeader(jwsHeader));
}
```

I'm using:
- spring boot and `spring-boot-starter-oauth2-resource-server ` 2.6.3
- It's using `ReactiveJwtDecoder` because I'm building an app with Spring Webflux

Not sure what I can do to fix this... 🤔

**Comment by dhaeb** [comment]
@Blockost how did you deal with this issue? We created our own `NimbusReactiveJwtDecoder` but are not happy to maintain this code.
It seems like that: 
`NimbusJwtDecoder.withJwkSetUri`can handle `JWE`, the reactive version `NimbusReactiveJwtDecoder.withJwkSetUri`  does not work for us...

**Pull request merged**